### PR TITLE
perf(sft): drop multi-turn renderer splits that have zero training weight

### PR DESCRIPTION
**Pull request by benjibc**
## Summary

`_render_one_worker` already filters rendered datums by token length
(`2 <= len <= max_seq_len`). It does **not** filter out datums whose
`token_weights` are all zero. On datasets with per-message weights
(SFT teacher-force, mixed-supervision conversations, RLOR-style teacher
mimicking), the multi-turn renderer split emits examples whose target
assistant turn has `weight=0` — those examples are pure context for
the prefix and produce zero gradient, but still cost a full forward
pass.

## Measured impact

On one such dataset (Kimi K2.5/K2.6 chat-with-tools, mixed weights):

| Stat | Value |
|---|---|
| Assistant turns with `weight=0` | **73.5 %** |
| Rows with >1 user turn (renderer splits) | 66.1 % |
| Renderer-emitted splits that target a `weight=0` assistant | **88.1 %** |
| Per-step tokens shipped to the trainer that produced zero loss | **~73 %** |

## Fix

A single condition added to the existing datum filter:

```python
datums = [
    rendered.datum
    for rendered in rendered_examples
    if 2 <= len(rendered.token_ids) <= _worker_state["max_seq_len"]
    and any(w != 0 for w in rendered.token_weights)   # NEW
]
```

If a rendered example has zero training weight across all positions,
drop it before the trainer ever sees it.

## Notes

- Same code path is dispatched by managed SFT V2
  (`training_orchestrator._run_sft` → `training.recipes.sft_loop.main`),
  so the perf win flows through to every customer SFT job automatically.
- A more thorough fix would prune zero-loss user-turn boundaries inside
  `tinker_cookbook`'s `build_supervised_examples` so they're never
  rendered in the first place (saves CPU on the dataloader side too).
  This cookbook-side filter is the smallest diff that captures the
  GPU-side win.

---

> [!NOTE]
> **Medium Risk**
> Changes SFT dataset filtering to skip examples with zero training weight, which can alter effective training data/step counts on weighted datasets if weights are mis-specified or all-zero.
> 
> **Overview**
> **SFT rendering now drops zero-loss examples.** `_render_one_worker` filters out renderer-emitted datums whose `token_weights` are entirely non-positive, in addition to the existing sequence-length bounds, to avoid wasting forward passes on examples that contribute no gradient.
> 
> Unit tests update their `render_messages_to_datums` stubs to include `token_weights` so the new filter path is exercised and existing training-loop behaviors remain covered.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7029864e2b67f20e26957c091fe760eebbcb8ae9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by websterbei — APPROVED** [review]

File: training/recipes/sft_loop.py [hunk 1]
```diff
@@ -114,6 +114,7 @@ def _render_one_worker(row: dict) -> tinker.Datum | list[tinker.Datum] | None:
         rendered.datum
         for rendered in rendered_examples
         if 2 <= len(rendered.token_ids) <= _worker_state["max_seq_len"]
+        and any(w != 0 for w in rendered.token_weights)
```
Comment (websterbei) [inline training/recipes/sft_loop.py]: it could be float

Comment (websterbei) [reply training/recipes/sft_loop.py]: maybe safer using >= 0

**Review by websterbei — APPROVED** [review]

**Comment by benjibc** [comment]
Updated to use `w > 0` (`292f0c3`).

Reasoning: weights here are floats. `w > 0` is semantically clearer ("has positive training signal") and defensive against any potential negative-weight or tiny-float edge case while keeping the behavior on `weight == 0` datums (drop) unchanged. `>= 0` would also keep `weight == 0` and defeat the filter, so I read your comment as "switch from `!= 0` to a positive-signal check" — let me know if you meant something different.

File: training/recipes/sft_loop.py (hunk 1 above)
Comment (websterbei) [reply training/recipes/sft_loop.py]: ^ the first comment is not true. I'm forgetting Python syntax

**Review by xiaoyifan — APPROVED** [review]
